### PR TITLE
Frontend session stuff

### PR DESCRIPTION
**Pull request by fhartwig**
Not ready to merge yet.

### CODE REVIEW

File: src/Thentos/Frontend.hs [hunk 1]
```diff
@@ -208,6 +206,48 @@ loginHandler = do
...
+
+I think this may require more fundamental changes to modules Thentos.Api.*, unless you make checkPassword a transaction in Thentos.DB.Trans. Currently, transactions are not only the thing that carries atomicity, but also the thing that carries labels. Actions carry neither. Probably something we will need to fix eventually.
+-}
+            case emUser of
+                Right (Just (uid, _)) -> with sess $ do
+                    setInSession "user" (cs $ Aeson.encode [uid])
+                    commitSession
+                    blaze "Logged in"
+                Right Nothing -> loginFail
+                Left NoSuchUser -> loginFail
```
Comment (mf59816) [inline src/Thentos/Frontend.hs]: This line should be deleted now, together with the FIXME.

File: src/Thentos/Frontend.hs [hunk 2]
```diff
@@ -208,6 +206,48 @@ loginHandler = do
...
+logIntoThentosHandler = do
+    (_view, result) <- runForm "log_into_thentos" loginForm
+    case result of
+        Just (username, password) -> do
+            emUser <- snapRunAction' allowEverything $ checkPassword username password
+{-
+checkPassword should not require allowEverything, because the context in which this is called is usually without any clearance. So allowEverything should be allowNothing, and checkPassword should be labelled thentosPublic.
+
+I think this may require more fundamental changes to modules Thentos.Api.*, unless you make checkPassword a transaction in Thentos.DB.Trans. Currently, transactions are not only the thing that carries atomicity, but also the thing that carries labels. Actions carry neither. Probably something we will need to fix eventually.
+-}
```
Comment (mf59816) [inline src/Thentos/Frontend.hs]: With #20, this can be fixed now.

File: src/Thentos/Frontend.hs (hunk 1 above)
Comment (mf59816) [inline src/Thentos/Frontend.hs]: (There is at least one more instance of the same oversight.)

File: services/helloworld/src/Site.hs [hunk 3]
```diff
@@ -114,18 +120,37 @@ tokenOk :: Maybe ByteString -> Handler App App Bool
 tokenOk Nothing = return False
 tokenOk (Just token) = do
     hwConfig <- gets aHWConfig
-    let sid = serviceId hwConfig
-        key = serviceKey hwConfig
-        url = thentosBackendUrl hwConfig <> "/session/" <> urlEncode token <> "/login/" <> urlEncode sid
+    let sid = encodeUtf8 $ serviceId hwConfig
```
Comment (mf59816) [inline services/helloworld/src/Site.hs]: change type in hwConfig rather than re-encode here.

File: services/helloworld/src/Site.hs [hunk 4]
```diff
@@ -114,18 +120,37 @@ tokenOk :: Maybe ByteString -> Handler App App Bool
...
+        url = thentosBackendUrl hwConfig <> "/session"
+        json_tok = "{\"fromSessionToken\":\"" <> token <> "\"}"
```
Comment (mf59816) [inline services/helloworld/src/Site.hs]: I would prefer:

    Aeson.object ["fromSessionToken" .= token]

and then in the next line:

    reqBody = RequestBodyBS . cs . Aeson.encode $ json_tok

File: services/helloworld/src/Site.hs [hunk 5]
```diff
@@ -114,18 +120,37 @@ tokenOk :: Maybe ByteString -> Handler App App Bool
...
+getServiceSessionToken :: Handler App App ByteString
+getServiceSessionToken = do
+    hwConfig <- gets aHWConfig
+    let sid = serviceId hwConfig
+        key = serviceKey hwConfig
+        url = thentosBackendUrl hwConfig <> "/session"
+        json_sid = "{\"fromServiceId\":" <> Aeson.encode sid <> "}"
+        json_key = "{\"fromServiceKey\":" <> Aeson.encode key <> "}"
+        json = "[" <> json_sid <> ", " <> json_key <> "]"
+        reqBody = RequestBodyLBS $ json
```
Comment (mf59816) [inline services/helloworld/src/Site.hs]: see last comment.  (here that's even more of an improvement.)

**Comment by mf59816** [comment]
Looks good,except: (1) above comments; (2) UI still incomplete, so it's a bit awkward to test.

Please address (1) and re-assign back to em.  (2) can go on a different branch.

File: services/helloworld/src/Site.hs (hunk 3 above)
Comment (fhartwig) [inline services/helloworld/src/Site.hs]: I'd rather not. If I did that, than I'd have to `decodeUtf8` it in a bunch of places. At least `encodeUtf8` is total.

File: services/helloworld/src/Site.hs (hunk 4 above)
Comment (fhartwig) [inline services/helloworld/src/Site.hs]: Yes, that's better.

File: src/Thentos/Frontend.hs (hunk 1 above)
Comment (fhartwig) [inline src/Thentos/Frontend.hs]: Well-spotted. Will fix.

File: services/helloworld/src/Site.hs (hunk 3 above)
Comment (mf59816) [inline services/helloworld/src/Site.hs]: Ok, you're right.  (s/than/then/, though.  :-)

**Pull request merged**